src/vulkan/vulkan_renderer.cpp

```diff
@@ -296,24 +312,39 @@ namespace IC {
         }
 
         // material descriptors (set 1 or 2)
-        WriteMaterialDescriptors(_allocator, SwapChain::MAX_FRAMES_IN_FLIGHT, writer, *materialData, _textureManager,
+        WriteMaterialDescriptors(_allocator, SwapChain::MAX_FRAMES_IN_FLIGHT, writer, *mesh.Material(), _textureManager,
                                  meshRenderData.materialBuffers);
         for (size_t i = 0; i < SwapChain::MAX_FRAMES_IN_FLIGHT; i++) {
-            int index = materialData->Template().flags & MaterialFlags::Lit ? 2 : 1;
+            int index = mesh.Material()->Template().flags & MaterialFlags::Lit ? 2 : 1;
             writer.UpdateSet(_vulkanDevice.Device(), meshRenderData.descriptorSets[i][index]);
             meshRenderData.UpdateMaterialBuffer(i);
         }
 
         _renderData.push_back(meshRenderData);
     }
 
-    void VulkanRenderer::AddLight(std::shared_ptr<PointLight> light) {
-        _pointLights.push_back(light);
-        AddMesh(light->previewMesh, light->previewMaterial);
+    void VulkanRenderer::AddDirectionalLight(std::shared_ptr<DirectionalLight> &light) {
+        _lightData.directionalLight = light;
+    }
+
+    void VulkanRenderer::AddPointLight(std::shared_ptr<PointLight> &light, std::shared_ptr<Transform> &transform) {
+        if (_lightData.pointLights.size() < MAX_POINT_LIGHTS) {
+            _lightData.pointLights.push_back({light, transform});
+        }
     }
 
-    void VulkanRenderer::AddDirectionalLight(std::shared_ptr<DirectionalLight> light) {
-        _directionalLight = light;
+    void VulkanRenderer::AddGameObject(std::shared_ptr<GameObject> object) {
+        if (object->HasComponent<Mesh>()) {
+            AddMesh(*object->GetComponent<Mesh>(), *object->GetTransform());
+        }
+        if (object->HasComponent<PointLight>()) {
+            auto light = object->GetComponent<PointLight>();
+            AddPointLight(light, object->GetTransform());
```

Comment:
here I had to assign a variable instead of directly putting `object->GetComponent<PointLight>()` in `AddPointLight()` otherwise I got this error `cannot bind non-const lvalue reference of type ‘std::shared_ptr<IC::PointLight>&’ to an rvalue of type ‘std::shared_ptr<IC::PointLight>’`

Reply:
Ah yes, so this happens because the function is expecting a non-const reference of the object, but the best the function return value can do is a const reference. It would work if you made the first parameter of the `AddPointLight` function a const reference, assuming the function doesn't try to modify the shared_ptr in any way. But you don't have to, just wanted to explain why this is happening.